File: scala/revenj-akka/src/main/scala/net/revenj/server/commands/search/CheckDomainObject.scala
package net.revenj.server.commands.search

import net.revenj.patterns._
import net.revenj.serialization.Serialization
import net.revenj.server.commands.search.CheckDomainObject.Argument
import net.revenj.server.{CommandResult, ReadOnlyServerCommand}

import scala.concurrent.Future

class CheckDomainObject(domainModel: DomainModel) extends ReadOnlyServerCommand {

  override def execute[TInput, TOutput](
    locator: ServiceLocator,
    input: Serialization[TInput],
    output: Serialization[TOutput],
    data: TInput): Future[CommandResult[TOutput]] = {

    val arg = input.deserialize[Argument](data)
    lazy val manifest = domainModel.find(arg.get.Name)
    if (!arg.isSuccess) {
      CommandResult.badRequest(arg.failed)
    } else if (manifest.isEmpty) {
      CommandResult.badRequest(s"Unable to find specified domain object: ${arg.get.Name}")
    } else if (arg.get.Uri == null) {
      CommandResult.badRequest("Uri to find not specified.")
    } else if (!classOf[Identifiable].isAssignableFrom(manifest.get)) {
      CommandResult.badRequest(s"Specified type is not an identifiable: ${arg.get.Name}")
    } else {
      val tryRepository = locator.resolve(classOf[Repository[Identifiable]], manifest.get)
      if (!tryRepository.isSuccess) {
        CommandResult.badRequest(s"Error resolving repository for: ${arg.get.Name}. Reason: ${tryRepository.failed.get.getMessage}")
      } else {
        import scala.concurrent.ExecutionContext.Implicits.global
        tryRepository.get.find(Array(arg.get.Uri)).map { found =>
          val response = output.serializeRuntime(found.length == 1)
          if (response.isSuccess) {
            CommandResult[TOutput](Some(response.get), if (found.length == 1) "Found" else "Not found", 200)
          } else {
            CommandResult[TOutput](None, response.failed.get.getMessage, 500)
          }
        }
      }
    }
  }
}

object CheckDomainObject {
  case class Argument(Name: String, Uri: String)
}

File: java/revenj-servlet/src/main/java/org/revenj/server/commands/search/CheckDomainObject.java
package org.revenj.server.commands.search;

import com.dslplatform.json.CompiledJson;
import com.dslplatform.json.JsonAttribute;
import org.revenj.patterns.*;
import org.revenj.security.PermissionManager;
import org.revenj.serialization.Serialization;
import org.revenj.server.CommandResult;
import org.revenj.server.ReadOnlyServerCommand;

import java.io.IOException;
import java.lang.reflect.Type;
import java.security.Principal;
import java.util.Optional;

public class CheckDomainObject implements ReadOnlyServerCommand {

	private final DomainModel domainModel;
	private final PermissionManager permissions;

	public CheckDomainObject(
			DomainModel domainModel,
			PermissionManager permissions) {
		this.domainModel = domainModel;
		this.permissions = permissions;
	}

	@CompiledJson
	public static final class Argument {
		@JsonAttribute(name = "Name", alternativeNames = {"name"})
		public final String name;
		@JsonAttribute(name = "Uri", alternativeNames = {"uri"})
		public final String uri;

		public Argument(String name, String uri) {
			this.name = name;
			this.uri = uri;
		}
	}

	@Override
	public <TInput, TOutput> CommandResult<TOutput> execute(
			ServiceLocator locator,
			Serialization<TInput> input,
			Serialization<TOutput> output,
			TInput data,
			Principal principal) {
		Argument arg;
		try {
			arg = input.deserialize(data, Argument.class);
		} catch (IOException e) {
			return CommandResult.badRequest(e.getMessage());
		}
		Optional<Class<?>> manifest = domainModel.find(arg.name);
		if (!manifest.isPresent()) {
			return CommandResult.badRequest("Unable to find specified domain object: " + arg.name);
		}
		if (arg.uri == null) {
			return CommandResult.badRequest("Uri not specified.");
		}
		if (!Identifiable.class.isAssignableFrom(manifest.get())) {
			return CommandResult.badRequest("Specified type is not an identifiable: " + arg.name);
		}
		if (!permissions.canAccess(manifest.get(), principal)) {
			return CommandResult.forbidden(arg.name);
		}
		Repository repository;
		try {
			repository = locator.resolve(Repository.class, manifest.get());
		} catch (ReflectiveOperationException e) {
			return CommandResult.badRequest("Error resolving repository for: " + arg.name + ". Reason: " + e.getMessage());
		}
		Optional<AggregateRoot> found = repository.find(arg.uri);
		try {
			return new CommandResult<>(output.serialize(found.isPresent()), found.isPresent() ? "Found" : "Not found", 200);
		} catch (IOException e) {
			return new CommandResult<>(null, "Error serializing result.", 500);
		}
	}
}
